src/adapter/src/coord/timestamp_oracle/postgres.rs

```diff
@@ -0,0 +1,399 @@
+
+        let q = r#"
+            UPDATE timestamp_oracle SET write_ts = GREATEST(write_ts+1, $2)
+                WHERE timeline = $1
+            RETURNING write_ts;
+        "#;
+        let client = self.get_connection().await.expect("todo");
+        let statement = client.prepare_cached(q).await.expect("todo");
+        let rows = client
+            .query(&statement, &[&self.timeline, &proposed_next_ts])
+            .await
+            .expect("todo");
+
+        assert!(rows.len() == 1);
+
+        let result = rows.first().unwrap();
+
+        let write_ts: i64 = result.try_get("write_ts").expect("todo");
+        let write_ts = Timestamp::try_from(write_ts).expect("todo");
+
+        debug!(write_ts = ?write_ts, "write_ts");
+
+        let advance_to = write_ts.step_forward();
+
+        WriteTimestamp {
+            timestamp: write_ts,
+            advance_to,
+        }
+    }
+
+    #[tracing::instrument(name = "oracle::peek_write_ts", level = "debug", skip_all)]
+    async fn peek_write_ts(&self) -> Timestamp {
+        let q = r#"
+            SELECT write_ts FROM timestamp_oracle
+                WHERE timeline = $1;
+        "#;
+        let client = self.get_connection().await.expect("todo");
+        let statement = client.prepare_cached(q).await.expect("todo");
+        let rows = client
+            .query(&statement, &[&self.timeline])
+            .await
+            .expect("todo");
+
+        assert!(rows.len() == 1);
+
+        let result = rows.first().unwrap();
+
+        let write_ts: i64 = result.try_get("write_ts").expect("todo");
+        let write_ts = Timestamp::try_from(write_ts).expect("todo");
+
+        debug!(write_ts = ?write_ts, "peek_write_ts");
+
+        write_ts
+    }
+
+    #[tracing::instrument(name = "oracle::read_ts", level = "debug", skip_all)]
+    async fn read_ts(&self) -> Timestamp {
+        let q = r#"
+            SELECT read_ts FROM timestamp_oracle
+                WHERE timeline = $1;
+        "#;
+        let client = self.get_connection().await.expect("todo");
+        let statement = client.prepare_cached(q).await.expect("todo");
+        let rows = client
+            .query(&statement, &[&self.timeline])
+            .await
+            .expect("todo");
+
+        assert!(rows.len() == 1);
+
+        let result = rows.first().unwrap();
+
+        let read_ts: i64 = result.try_get("read_ts").expect("todo");
+        let read_ts = Timestamp::try_from(read_ts).expect("todo");
+
+        debug!(read_ts = ?read_ts, "read_ts");
+
+        read_ts
+    }
+
+    #[tracing::instrument(name = "oracle::apply_write", level = "debug", skip_all)]
+    async fn apply_write(&mut self, write_ts: Timestamp) {
+        let q = r#"
+            UPDATE timestamp_oracle SET write_ts = GREATEST(write_ts, $2), read_ts = GREATEST(read_ts, $2)
+                WHERE timeline = $1;
+        "#;
+        let client = self.get_connection().await.expect("todo");
```

Comment:
This is being seen in Sentry - https://materializeinc.sentry.io/issues/4478589600/?project=6780145&query=is%3Aunresolved&referrer=issue-stream&statsPeriod=7d&stream_index=10